cpp/src/neighbors/cagra.cuh

```diff
@@ -394,6 +395,188 @@ void search(raft::resources const& res,
+ */
+template <class T, class IdxT, class Accessor>
+void add_graph_nodes(
+  raft::resources const& handle,
+  raft::mdspan<const T, raft::matrix_extent<int64_t>, raft::layout_stride, Accessor>
+    input_updated_dataset_view,
+  const neighbors::cagra::index<T, IdxT>& index,
+  raft::host_matrix_view<IdxT, std::int64_t> updated_graph_view,
+  const std::size_t max_batch_size)
+{
+  assert(input_updated_dataset_view.extent(0) >= index.size());
+
+  const std::size_t initial_dataset_size = index.size();
+  const std::size_t new_dataset_size     = input_updated_dataset_view.extent(0);
+  const std::size_t num_new_nodes        = new_dataset_size - initial_dataset_size;
+  const std::size_t degree               = index.graph_degree();
+  const std::size_t dim                  = index.dim();
+  const std::size_t stride               = input_updated_dataset_view.stride(0);
+  const std::size_t max_batch_size_      = max_batch_size == 0 ? 1 : max_batch_size;
+
+  raft::copy(updated_graph_view.data_handle(),
+             index.graph().data_handle(),
+             initial_dataset_size * degree,
+             raft::resource::get_cuda_stream(handle));
+
+  neighbors::cagra::index<T, IdxT> internal_index(
+    handle,
+    index.metric(),
+    raft::make_device_matrix_view<const T, int64_t>(nullptr, 0, 0),
+    raft::make_device_matrix_view<const IdxT, int64_t>(nullptr, 0, 0));
+
+  for (std::size_t additional_dataset_offset = 0; additional_dataset_offset < num_new_nodes;
+       additional_dataset_offset += max_batch_size_) {
+    const auto actual_batch_size =
+      std::min(num_new_nodes - additional_dataset_offset, max_batch_size_);
+
+    auto dataset_view = raft::make_device_strided_matrix_view<const T, std::int64_t>(
+      input_updated_dataset_view.data_handle(),
+      initial_dataset_size + additional_dataset_offset,
+      dim,
+      stride);
+    auto graph_view = raft::make_host_matrix_view<const IdxT, std::int64_t>(
+      updated_graph_view.data_handle(), initial_dataset_size + additional_dataset_offset, degree);
+
+    internal_index.update_dataset(handle, dataset_view);
+    // Note: The graph is copied to the device memory.
+    internal_index.update_graph(handle, graph_view);
+    raft::resource::sync_stream(handle);
+
+    auto updated_graph = raft::make_host_matrix_view<IdxT, std::int64_t>(
+      updated_graph_view.data_handle(),
+      initial_dataset_size + additional_dataset_offset + actual_batch_size,
+      degree);
+    auto additional_dataset_view = raft::make_device_strided_matrix_view<const T, std::int64_t>(
+      input_updated_dataset_view.data_handle() +
+        (initial_dataset_size + additional_dataset_offset) * stride,
+      actual_batch_size,
+      dim,
+      stride);
+
+    neighbors::cagra::detail::add_node_core<T, IdxT>(
+      handle, internal_index, additional_dataset_view, updated_graph);
+    raft::resource::sync_stream(handle);
+  }
+}
+
+/**
+ * @brief Add new vectors to a CAGRA index
```

Comment:
Perhaps this comment should be moved to the hpp file because that is what is going to be visible to end users.

Reply:
Is it better to move the other comments to the hpp in this PR? or separate PR?

Reply:
I only moved the comment of the functions available in the include/hpp file. Some functions in the cuh file are not in the hpp file, and I didn't move their comments. I'll make a separate PR to make them public APIs and port the tests for them from RAFT for the next release.